Restructure Build apps: add an Arbitrum essentials section (#3332)
OffchainLabs/arbitrum-docs: docs/arbitrum-essentials/bridging/configure-token-gateway/custom.mdx

user_story: 'As a developer, I want to deploy a custom token bridge gateway so I can implement specialized bridging logic for my token between Ethereum and Arbitrum.'
reader_audience: developers who want to build on Ethereum/Arbitrum and bridge tokens between parent and child chains
content_type: how-to
displayed_sidebar: arbitrumEssentialsSidebar
---

:::caution Do you really need a custom gateway?

> Before implementing and deploying a custom gateway, it is strongly encouraged to analyze the current solutions that Arbitrum's token bridge provides: the [standard gateway](/arbitrum-essentials/bridging/configure-token-gateway/standard.mdx) and the [generic-custom gateway](/arbitrum-essentials/bridging/configure-token-gateway/generic-custom.mdx). These solutions provide enough functionality to solve the majority of bridging needs from projects. And if you are in doubt about your current approach, you can always ask for assistance on our [Discord server](https://discord.gg/arbitrum).

Author: ```suggestion
Before implementing and deploying a custom gateway, strongly consider the current solutions that Arbitrum's token bridge provides: the [standard gateway](/arbitrum-essentials/bridging/configure-token-gateway/standard.mdx) and the [generic-custom gateway](/arbitrum-essentials/bridging/configure-token-gateway/generic-custom.mdx). These solutions cover the majority of bridging needs. If you're unsure about your approach, you can always ask for assistance on our [Discord server](https://discord.gg/arbitrum).
```

:::

In this how-to, you'll learn how to bridge your own token between Ethereum (the parent chain) and Arbitrum (the child chain), using a custom gateway. For alternative ways of bridging tokens, check out the [token bridging overview](/arbitrum-essentials/bridging/overview.mdx).

Familiarity with [Arbitrum's token bridge system](/how-arbitrum-works/deep-dives/token-bridging.mdx), smart contracts, and decentralized application development is expected. If you're new to developing on Arbitrum, consider reviewing our [Quickstart: Build a dApp with Arbitrum (Solidity, Remix)](/build-decentralized-apps/01-quickstart-solidity-remix.mdx) before proceeding. We'll use [Arbitrum's SDK](https://github.com/OffchainLabs/arbitrum-sdk) throughout this how-to, although no prior knowledge is required.

We will go through all the steps involved in the process. However, if you want to jump straight to the code, we have created a [custom gateway bridging tutorial script](https://github.com/OffchainLabs/arbitrum-tutorials/tree/master/packages/custom-gateway-bridging) that encapsulates the entire process.

## Step 0: Review the prerequisites (a.k.a. do I really need a custom gateway?)

> Before implementing and deploying a custom gateway, it is strongly encouraged to analyze the current solutions that Arbitrum’s token bridge provides: the [standard gateway](/arbitrum-essentials/bridging/deposit/tokens.mdx) and the [generic-custom gateway](/arbitrum-essentials/bridging/configure-token-gateway/generic-custom.mdx). These solutions provide enough functionality to solve the majority of bridging needs from projects. And if you are in doubt about your current approach, you can always ask for assistance on our [Discord server](https://discord.gg/arbitrum).

Author: ```suggestion
Before implementing and deploying a custom gateway, strongly consider the current solutions that Arbitrum's token bridge provides: the [standard gateway](/arbitrum-essentials/bridging/configure-token-gateway/standard.mdx) and the [generic-custom gateway](/arbitrum-essentials/bridging/configure-token-gateway/generic-custom.mdx). These solutions cover the majority of bridging needs. If you're unsure about your approach, you can always ask for assistance on our [Discord server](https://discord.gg/arbitrum).
```

- A method `outboundTransfer`, to handle external calls, and forwarded calls from `L2GatewayRouter.outboundTransfer`.
- A method `finalizeInboundTransfer`, to handle messages coming **only** from the parent chain's gateway.
- Two methods, `calculateL2TokenAddress` and `getOutboundCalldata`, to handle other bridging operations.
- Methods to send cross-chain messages through the [ArbSys precompile](/arbitrum-essentials/precompiles/reference.mdx#arbsys). You can view an example implementation of `sendTxToL1` on [L2ArbitrumMessenger](https://github.com/OffchainLabs/token-bridge-contracts/blob/main/contracts/tokenbridge/arbitrum/L2ArbitrumMessenger.sol).

### What about my custom tokens?

> If you are deploying custom gateways, you will likely want to support your custom tokens on both the parent and child chains. They also have several requirements they must comply with. You can find more information about it in [How to bridge tokens via Arbitrum’s generic-custom gateway](/arbitrum-essentials/bridging/configure-token-gateway/generic-custom.mdx).

Author: ```suggestion
If you are deploying custom gateways, you will likely want to support your custom tokens on both the parent and child chains. They must also meet several requirements, detailed in [How to bridge tokens via Arbitrum's generic-custom gateway](/arbitrum-essentials/bridging/configure-token-gateway/generic-custom.mdx).
```

This step will depend on your setup. In this case, since our simplified gateway supports only one token, we'll deploy it on both the parent and child chains to enable calling the `setTokenBridgeInformation` method on both gateways afterwards.

> We won't go through the process of deploying custom tokens in this how-to, but you can see a detailed explanation of the steps to take on the [generic-custom gateway setup page](/arbitrum-essentials/bridging/configure-token-gateway/generic-custom.mdx).

Author: ```suggestion
We won't go through deploying custom tokens in this how-to, but you can find a detailed explanation on the [generic-custom gateway setup page](/arbitrum-essentials/bridging/configure-token-gateway/generic-custom.mdx).
```

Once all contracts are deployed successfully on their respective chains, and they all have the information of the gateways and tokens, it's time to register the token in your custom gateway.

> As mentioned in [How to bridge tokens via Arbitrum’s generic-custom gateway](/arbitrum-essentials/bridging/configure-token-gateway/generic-custom.mdx), this action needs to be completed by the parent chain token, and we’ve implemented the function `registerTokenOnL2` to do it. So now we only need to call that function.

Author: ```suggestion
As mentioned in [How to bridge tokens via Arbitrum's generic-custom gateway](/arbitrum-essentials/bridging/configure-token-gateway/generic-custom.mdx), this action needs to be completed by the parent chain token, and we've implemented the function `registerTokenOnL2` to do it. Now we only need to call that function.
```
